### PR TITLE
Adding FP8 support through Native XLA Support

### PR DESCRIPTION
**Pull request by abhinavgoel95**
@rwitten, following the discussion on https://github.com/google/maxtext/pull/468, I have made changes to call FP8 GeMMs in MaxText through the native XLA support.

I have tested the performance, and looks good on NVIDIA H100.

### CODE REVIEW

File: MaxText/configs/base.yml
Comment (rwitten) [inline MaxText/configs/base.yml:63]: nit: Seems like this should be quantization=fp8 two lines above? Feel free to push back.

**Review by rwitten — COMMENTED** [review]
Looks really promising. I left one nit.

I want Anfal on Google's side to take the first look, I will make sure he responds super prompotly.

File: MaxText/layers/linears.py
Comment (anfals) [inline MaxText/layers/linears.py:109]: Functionally, this looks great!

But, I think it may be cleaner if we added some conditional logic here instead `https://github.com/google/maxtext/blob/main/MaxText/layers/quantizations.py#L95`. We could introduce a general Quantization class (and existing AqtQuantization could subclass). We could then have an FP8 implementation of that class that returns nn.Fp8DotGeneralOp when dot_general_cls() is called. That is a little bit more work but feels a bit better to me. But open to hear other thoughts.

File: MaxText/max_utils.py
Comment (anfals) [inline MaxText/max_utils.py:353]: Ah is this because the flax linen impl adds those variables in the `overwrite_with_gradients` namespace?

File: MaxText/train.py
Comment (anfals) [inline MaxText/train.py:220]: same question as above :)

File: MaxText/max_utils.py
Comment (abhinavgoel95) [reply MaxText/max_utils.py:353]: Yes.

File: MaxText/train.py
Comment (abhinavgoel95) [reply MaxText/train.py:220]: Yes, same answer.

File: MaxText/configs/base.yml
Comment (abhinavgoel95) [reply MaxText/configs/base.yml:63]: Okay, I can dix that.

File: MaxText/layers/linears.py
Comment (abhinavgoel95) [reply MaxText/layers/linears.py]: Made some changes. I didn't know what the general Quantization class should look like. So I've omitted it for now. Lmk if you'd still like me to introduce it.

File: MaxText/configs/base.yml
Comment (abhinavgoel95) [reply MaxText/configs/base.yml]: Done this. @rwitten @anfals

File: MaxText/layers/quantizations.py
Comment (anfals) [inline MaxText/layers/quantizations.py:52]: This is great! Maybe we could have some class that looks like as the parent for the two? More of a nit
```
@dataclass
class Quantization:
    """Base class for quantization configurations"""

    def dot_general_cls(self):
        """ Placeholder for dot_general implementation in subclasses. """
        pass 
```

**Review by anfals — COMMENTED** [review]
Looks great! Thanks for this! I left one minor nit comment

File: MaxText/layers/linears.py
Comment (anfals) [reply MaxText/layers/linears.py]: Awesome! Thanks for this. I've left a comment about what I think it could look like but let me know your thoughts!

File: MaxText/layers/quantizations.py
Comment (abhinavgoel95) [reply MaxText/layers/quantizations.py:52]: Cool, Ill do that right away

Comment (abhinavgoel95) [reply MaxText/layers/quantizations.py:60]: Done, added what you suggested.

**Review by anfals — APPROVED** [review]

**Review by anfals — APPROVED** [review]

File: MaxText/train.py
Comment (rwitten) [inline MaxText/train.py:214]: I don't understand this forking. It is because of the difference between passing in `params` and {'params' : params}? Let's find a solution without that forking

File: MaxText/max_utils.py
Comment (rwitten) [inline MaxText/max_utils.py:352]: This is very suspicious to me. Why is it necessary, why can't we find a more universal solution?

**Review by rwitten — CHANGES_REQUESTED** [review]
Some code smell, happy to help folks understand it.

File: MaxText/max_utils.py
Comment (anfals) [reply MaxText/max_utils.py:352]: > Ah is this because the flax linen impl adds those variables in the overwrite_with_gradients namespace?

@rwitten See the resolved conversation below. `fp8ops` has the "hack" to have some variables keyed under `overwrite_with_gradients` so they can be manually overwritten, which leads to this weirdness. But agree - its odd. Maybe there is a better way to do this...

Comment (anfals) [reply MaxText/max_utils.py:352]: @abhinavgoel95 Hmm what happens if we just pass in `model_vars` always here? Looks like before in MaxText we were extracting the params directly, but Flax should be able to just infer the direct parameter collection as being equivalent to `{'params': params}`

Comment (abhinavgoel95) [reply MaxText/max_utils.py:352]: This is needed by the Fp8 implementation in Flax.

File: MaxText/train.py
Comment (abhinavgoel95) [reply MaxText/train.py:214]: cc @kaixih, do you know if there is a way around this?

File: MaxText/max_utils.py
Comment (abhinavgoel95) [reply MaxText/max_utils.py:352]: cc @kaixih

File: MaxText/train.py
Comment (anfals) [reply MaxText/train.py:214]: @abhinavgoel95 I'll also poke around with this a bit more today. I think there should be a way to just make this consistent throughout maxtext such that we just pass `params` around directly (so it'll always be {'params': params} by default). But that might take a little bit of extra work in this PR. I'll report back

Comment (kaixih) [reply MaxText/train.py:214]: The current behavior of TrainState is such that the pytree structure of optimizer states depends on what the user provides:

1. If the input is {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy}, the optimizer state will be {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy}.
2. If the input is {'w': xxx, 'b': xxx}, the optimizer state will be {'w': xxx, 'b': xxx}.
3. If the input is {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy, 'overwrite_with_gradient': zzz}, the optimizer state will be {'w': xxx, 'b': xxx}.

The behavior of (1) and (2) represents the optimizer states when we pass `model_var` and `model_var['params']` into TrainState. However, (3) passes the full model variables while the optimizer state only focuses on 'params' and disregards 'other_params', which might need some improvement if we do have any real-world use case that we need optimizer states for non 'params' variables.

Back to our scenario here, we are employing method (2) by originally passing `model_var['params']`. The proposed change entails utilizing (3) for fp8 configuration by passing the entire `model_vars`.

Considering this, if there are no real instances where `other_params` necessitates optimizer states, I suggest modifying the TrainState behavior in Flax for (1) as follows:

(1) {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy} -> opt {'w': xxx, 'b': xxx}

With this adjustment, the TrainState behavior remains consistent: it will generate optimizer states only for 'params' if they exist (cases 1 and 3); otherwise, it will generate states for all items (case 2). Consequently, there's no need for branching here, and we can simply pass in the entire `model_vars`:
```python
init_training_state(model.apply, model_vars, tx)

```

Comment (anfals) [reply MaxText/train.py:214]: @kaixih Thank you for the thorough explanation! The behavior in (3) is strange to me - I don't understand why it differs from (1). I can raise this internally with the Flax team. That said, I think I might have a workaround for us here: 

@abhinavgoel95 For this PR, I did a little bit of debugging of MaxText (on a TPU, but I think this should generalize). Within `init_initial_state`, `model_vars` looks like the following: 

`{'params': {'token_embedder': {...}, 'decoder': {...}}}`

So we can just pass it directly to `init_training_state` directly. And then within `loss_fn` in `train.py`, we can just pass `params` along directly. With just those changes, I was able to get training running. And I believe it should generalize to your changes (I'll try to verify that today). 

That said, I think MaxText has been a little inconsistent with this params versus params[params], so I think we might get some test errors, so it might be take a bit of work to get this done properly.

Comment (anfals) [reply MaxText/train.py:214]: One other thing that would be good to have in this PR is some kind of additional smoke test for the fp8 case. Along with the `params` testing, that is increasing the scope of the PR, so it might be good to separate the two out into separate PRs (once we verify that the change I proposed actually helps) - 1 to switch just using params everywhere instead of params[parmas]) and then the fp8 work + test. If you have bandwidth/want to do both, great! If not, I'm happy to do the first PR and then you can incorporate the changes here.

Comment (hxssgaa) [reply MaxText/train.py:214]: @abhinavgoel95 I ran the code on H100 with FP8 quantization, but faced following issue.

```
flax.errors.ScopeCollectionNotFound: Tried to access "input_amax_history" from collection "_overwrite_with_gradient" in "/decoder/remat(scan(layers))/self_attention/query/Fp8DotGeneralOp_0" but the collection is empty. (https://flax.readthedocs.io/en/latest/api_reference/flax.errors.html#flax.errors.ScopeCollectionNotFound)
```

Is this error due to the discussion above?

Comment (anfals) [reply MaxText/train.py:214]: @hxssgaa That looks to be related to some errors related to scan layers. That will be fixed before this PR is merged. You can try setting `scan_layers` to False in the meantime

**Comment by abhinavgoel95** [comment]
@anfals @rwitten I have rebased the code on top of the latest changes. Looks cleaner now. Lmk what you think. Also works with scan = True.

**Review by anfals — CHANGES_REQUESTED** [review]
This is great! Thank you for all the work on this so far!

I left some comments with a few changes I recommend. In addition to those, I'd suggest the following: 

1) Decoding doesn't work yet with 'fp8' so I think we could put an assert somewhere like `decode.py` that prevents decoding when fp8 is enabled (which we can remove in the follow-up PR with decode support)
2.) We should include a test with this PR. Maybe the equivalent of `train_int8_smoke_test` for fp8?

Beyond that, I verified the latest changes in main gave us checkpointing support for the `_overwrite_with_gradients` collection for free for training. It's breaking in decoding for some reason, but there is other work that needs to be done there, so we can kick the can on that until the next PR.

**Comment by anfals** [comment]
> This is great! Thank you for all the work on this so far!
> 
> I left some comments with a few changes I recommend. In addition to those, I'd suggest the following:
> 
> 1. Decoding doesn't work yet with 'fp8' so I think we could put an assert somewhere like `decode.py` that prevents decoding when fp8 is enabled (which we can remove in the follow-up PR with decode support)
>    2.) We should include a test with this PR. Maybe the equivalent of `train_int8_smoke_test` for fp8?
> 
> Beyond that, I verified the latest changes in main gave us checkpointing support for the `_overwrite_with_gradients` collection for free for training. It's breaking in decoding for some reason, but there is other work that needs to be done there, so we can kick the can on that until the next PR.

Update on this: checkpointing 'works' with decoding, but the post-processing of the params in `load_params` within MaxEngine is very AQT-specific for quantization. We'll need to make some changes there to support the `_overwrite_with_gradients` collection with this FP8 quantization in the follow-up PR

File: MaxText/configs/base.yml
Comment (abhinavgoel95) [reply MaxText/configs/base.yml]: Done.

File: MaxText/train.py
Comment (abhinavgoel95) [reply MaxText/train.py]: Should be fixed now. Both with and without scan works fine.

**Review by anfals — APPROVED** [review]
LGTM, thanks!

File: .github/workflows/UnitTests.yml
Comment (anfals) [inline .github/workflows/UnitTests.yml:102]: We might wanna add this to the GPU tests below, but we can do that in the follow-up decode PR :). This should do a good job covering the functionality added!

**Review by anfals — APPROVED** [review]
LGTM, pending the tests passing!

**Review by rwitten — APPROVED** [review]

**Review by anfals — APPROVED** [review]

**Review by anfals — APPROVED** [review]

**Pull request merged**
